.github/workflows/sage.yml

```diff
@@ -76,40 +78,60 @@ jobs:
+          if [ -n "${{ env.REMOVE_PATCHES }}" ]
+          then
+            echo "(cd ../build/pkgs/${{ env.SPKG }}/patches rm -f ${{ env.REMOVE_PATCHES }}; :)" >> upstream/update-pkgs.sh
+          fi
+          echo "sed -i.bak \"/pushdef.*LT_VERSION/s/3[0-9.]*/4/\" ../build/pkgs/python3/spkg-configure.m4" >> upstream/update-pkgs.sh
+          echo "sed -i.bak \"/export.*proxy/d\" ../build/bin/sage-spkg" >> upstream/update-pkgs.sh
+          ls -l upstream
+      - name: System factors for pull requests
+        # See https://github.com/sagemath/sage/blob/develop/.github/workflows/docker.yml for available systems
+        id: pull-request-systems
+        if: ${{ github.event_name == 'pull_request' }}
+        run: >-
+          echo 'systems=[
+          "archlinux-latest",
+          "conda-forge",
+          ]' >> $GITHUB_OUTPUT
+      - name: System factors for push to main
+        # See https://github.com/sagemath/sage/blob/develop/.github/workflows/docker.yml for available systems
+        id: push-to-main-systems
+        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
+        run: >-
+          echo 'systems=[
+          "debian-bullseye",
+          "fedora-36",
+          "centos-stream-8-python3.9",
+          "gentoo-python3.9",
+          "archlinux-latest",
+          "opensuse-15.4-gcc_11-python3.10",
+          "opensuse-tumbleweed-python3.10",
+          "conda-forge",
+          "archlinux-latest",
+          "conda-forge",
+          ]' >> $GITHUB_OUTPUT
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
         with:
           path: upstream
           name: upstream
 
-  linux:
+  test:
     # https://github.com/sagemath/sage/blob/develop/.github/workflows/docker.yml
     # Use branch u/mkoeppe/numpy_1_23_x__scipy_1_9_x for a fix for uppercase github repo names (FFY00)
     uses: sagemath/sagetrac-mirror/.github/workflows/docker.yml@u/mkoeppe/numpy_1_23_x__scipy_1_9_x
     with:
-      # Reduced list of systems to test: Only those that bring python >= 3.8
-      # Full list at https://github.com/sagemath/sage/blob/develop/.github/workflows/docker.yml
-      # See also https://github.com/sagemath/sage/blob/develop/tox.ini
-      tox_system_factors: >-
-          ["debian-bullseye",
-           "fedora-36",
-           "centos-stream-8-python3.9",
-           "gentoo-python3.9",
-           "archlinux-latest",
-           "opensuse-15.4-gcc_11-python3.10",
-           "opensuse-tumbleweed-python3.10",
-           "conda-forge",
-           ]
+      tox_system_factors: ${{ needs.build.outputs.systems }}
       tox_packages_factors: >-
-          ["minimal-develop",
-           ]
-      # Extra packages to install as system packages
+        ["minimal-develop",]
+      # Extra system packages to install
       extra_sage_packages: "liblzma bzip2 libffi libpng python3 ninja_build"
```

Comment:
May want to add that these package names refer to https://github.com/sagemath/sage/tree/develop/build/pkgs

Reply:
Sure. By the way, do you know why we need these packages? I'm temped to remove them and see what happens.

Reply:
`python3` is needed because otherwise Sage will build its own copy of Python instead of the system python that is intended to be tested here.
`liblzma bzip2 libffi libpng` are needed because they are library dependencies of Python, and Sag will build its own copy of Python if they are not coming from the system.

Reply:
Also, do we still need to use your branch for the GitHub action or can we go back to the upstream one? What is the canonical source for it?

Reply:
See https://github.com/sagemath/sage/blob/develop/build/pkgs/python3/spkg-configure.m4#L23 (although I don't see libpng there)

Reply:
> `liblzma bzip2 libffi libpng` are needed because they are library dependencies of Python

Good to know. I thought that Sage did some dependency tracking. I'll add a comment.

Reply:
> (although I don't see libpng there)

Do we have a way to inspect if everything still works as intended if I remove `libpng` from the list?

Reply:
At the end of the `configure` run, you see lines such as this:
```
python3:                        using system package; SPKG will not be installed
```
https://github.com/mesonbuild/meson-python/actions/runs/3509068860/jobs/5877914584#step:11:1314

If the system python is no longer accepted, this line will change.

Reply:
> do we still need to use your branch for the GitHub action or can we go back to the upstream one?

Still need the branch; this will change it a few days.

Reply:
> > `liblzma bzip2 libffi libpng` are needed because they are library dependencies of Python
> 
> Good to know. I thought that Sage did some dependency tracking.

It does, but this goes in the opposite direction: If Sage does not find system `libffi` for example, it builds its own copy; and then it does not accept system `python3` (because that is bringing its own copy of `libffi`)

Reply:
I see. A quick note in passing: it would be nice for the Sage test action to upgrade to `actions/upload-artifact@v3` as the `v2` version emits a warning about `node` 12 going away soonish.

Reply:
`libpng` does not seem to be required indeed. Do you know if there is a way to cut down some of the time that goes into the image preparation? IIUC, there are is a lot of stuff that we don't need being searched for and compiled.

Reply:
> Do you know if there is a way to cut down some of the time that goes into the image preparation?

I've opened https://trac.sagemath.org/ticket/34765 to allow us to use system `meson`.

Reply:
I don't think that Meson is the biggest contender. Looking at the Arch job I see 224 SPKG packages installed.

Reply:
> upgrade to `actions/upload-artifact@v3` as the `v2` version emits a warning about `node` 12 going away soonish

Yes, I'll do this in https://trac.sagemath.org/ticket/34766

Reply:
> I don't think that Meson is the biggest contender. Looking at the Arch job I see 224 SPKG packages installed.

If you're referring to what `configure` reports as "will be installed" -- that's what would be installed if a full `make build` was done. You can see what is actually being installed in lines such as
https://github.com/mesonbuild/meson-python/actions/runs/3509347178/jobs/5878504732#step:11:1751
```
[meson-0.63.3] installing. Log file: /sage/logs/pkgs/meson-0.63.3.log
```

Reply:
I see. Thank you. I'm trying to understand what exactly in the setup phase takes 6 minutes for Arch and 14 minutes for Conda and see if there is a way to cut down this time.

Reply:
Just for context, a full build of Sage (`make build`) takes between 3 and 6 hours on GH Actions, depending on what system package configuration is used.

Reply:
Sure, but all we need is a regular distribution image with Python installed. Maybe the Sage test infrastructure is not the right place where to start...

Reply:
I have no objections if you want to replace the Sage CI by something that you maintain by yourself